bpo-29615: SimpleXMLRPCDispatcher no longer chains KeyError (or any other exception) to exception(s) raised in the dispatched methods.

petr-motejlek opened the pull request.

Commit:
issue29615: _dispatch function is no longer called in an except block

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow these steps to help rectify the issue:

1. If you don't have an account on b.p.o, please [create one](http://bugs.python.org/user?@template=register)
2. Make sure your GitHub username is listed in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/22867185/d3aedcf8-f137-11e6-8469-d5f7861a1073.png)
3. If you have not already done so, please sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/)
4. If you just signed the CLA, please **wait at least one US business day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
5. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

serhiy-storchaka commented on Lib/xmlrpc/server.py:
```diff
@@ -386,30 +386,31 @@ def _dispatch(self, method, params):
-        else:
-            raise Exception('method "%s" is not supported' % method)
+                def func(*params):
+                    return dispatch_func(method=method, params=params)
```
It is not required that the dispatch function supports keyword arguments.

What is the purpose of this change? Why not call the dispatch function directly?
```
return func(method, params)
```

petr-motejlek replied:
> It is not required that the dispatch function supports keyword arguments.
Agreed, changed.

> What is the purpose of this change? Why not call the dispatch function directly?
The purpose of this change is to make sure no exception handling is in progress when the function is called, thus I do attempt to first resolve the function (handling exceptions as I go) and then once the function is (hopefully) resolved, I attempt to call it

I did add `@functools.wraps` around the wrapper (forgot to check that in earlier)

serhiy-storchaka replied:
That is why I suggest to move all code after `except KeyError` out of exception handling block. This will allow to get rid of exception chaining and may make the code cleaner.

serhiy-storchaka commented on Lib/xmlrpc/server.py:
```diff
@@ -386,30 +386,31 @@ def _dispatch(self, method, params):
+                        self.allow_dotted_names
+                    )
+                except AttributeError:
+                    raise Exception('method "%s" is not supported' % method)
```
This chains two AttributeErrors and a KeyError to the raised exception.

petr-motejlek replied:
I know. But is that a problem, though? For me, this was the intended feature as it lets the caller know (using the exception chain) what steps were taken (and failed) to find the appropriate function

serhiy-storchaka replied:
That looks a problem for me. These exceptions don't add useful information and are implementation details. For example current implementation uses `hasattr()` rather than catching an `AttributeError`.

petr-motejlek replied:
Let's leave the current implementation aside in terms of what's correct and incorrect :)

However, I will remove the chaining, as you suggest.

petr-motejlek replied:
Chaining removed

serhiy-storchaka commented on Lib/xmlrpc/server.py:
```diff
@@ -386,30 +386,31 @@ def _dispatch(self, method, params):
+            except AttributeError:
+                # call instance method directly
+                try:
+                    func = resolve_dotted_attribute(
```
What if `resolve_dotted_attribute()` returns `None`?

petr-motejlek replied:
Hi @serhiy-storchaka , this is my 1st PR and 1st comment, so let me take this opportunity to thank you for the review :).

What about other (non-callable) objects, though? I mean, it's all the same and it doesn't really matter if whatever the function returns is `NoneType` or, let's say, `int(2)` --> it's still wrong. The original code does not check that either and it simply attempts to call the result. 

I would let Python deal with the NoneType and other non-callable types and raise it's own TypeError and leave this part of the code untouched. It is still a subclass of Exception, so the signature, if you will, of this function is still the same

serhiy-storchaka replied:
`None` is handled special in current code. Existing code can set an attribute to `None` for disabling an inherited method. A user get a more informative exception *'method "{methodname}" is not supported'* rather than *"'NoneType' object is not callable"*. I don't know whether this is used in real code, but there is a risk of breaking existing code.

petr-motejlek replied:
OK, let me fix that

petr-motejlek replied:
Fixed, although the code is now a little bit more bloated than I would have liked, but it works.

Commit:
issue29615 (review): use `functools.wraps` and remove the keyword-only argument call

Commit:
issue29615 (review): do not chain the exceptions, refrain from attempts to invoke `None`

serhiy-storchaka commented on Lib/xmlrpc/server.py:
```diff
@@ -103,7 +103,7 @@ def export_add(self, x, y):
 # Written by Brian Quinlan (brian@sweetapp.com).
 # Based on code written by Fredrik Lundh.
-
+import functools
```
Not used.

petr-motejlek replied:
Yup, gone.

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,6 +343,24 @@ def run_server():
             self.assertEqual(p.method(), 5)
             self.assertEqual(p.method(), 5)
 
+    def test_simple_xml_rpc_dispatcher_using_dispatch_func(self):
```
The test is passed when leave `Lib/xmlrpc/server.py` unchanged. Thus it doesn't test changes.

petr-motejlek replied:
My bad. I forgot to check in the proper test earlier.

serhiy-storchaka commented on Lib/xmlrpc/server.py:
```diff
@@ -387,29 +387,40 @@ def _dispatch(self, method, params):
+            except AttributeError:
+                pass
+
+            if func is not None:
```
For compatibility with current code it should be just `else` if use `try..except`.

serhiy-storchaka commented on Lib/xmlrpc/server.py:
```diff
@@ -387,29 +387,40 @@ def _dispatch(self, method, params):
+        if self.instance is not None:
+            try:
+                # check for a _dispatch method on the instance
+                func = self.instance._dispatch
```
It would be better to use other name for local variable.

But I suggest to keep the current code, with `hasattr()`. Catching exceptions is costly in Python, and if `_dispatch` is not defined this causes regression.

petr-motejlek replied:
Ok. Rewritten. I believe it is now in sync with the original code in every aspect, but the exception chaning issue I am attempting to solve.

serhiy-storchaka reviewed (changes requested):
Needed new test. Add an entry in Misc/NEWS.

Commit:
issue29615 (review): remove unused `functools`

Commit:
issue29615 (review): attempts to mimic the original behavior while getting rid of the exception chaining

Commit:
issue29615 (review): verifies KeyError`s (no exceptions actually) are no longer chained to exceptions raised within `_dispatch` function

Commit:
issue29615 (review): entry for these changes in Misc/NEWS

Commit:
Merge remote-tracking branch 'MASTER/master' into issue29615

Commit:
Merge remote-tracking branch 'MASTER/master' into issue29615

petr-motejlek commented:
Hi @serhiy-storchaka , in addition to the changes you requested directly in the code, I added a line to `Misc/NEWS` and merged the latest master. I hope we are getting closer to the end :)

EDIT: Travis complained about test coverage, so I added tests that ensure that all 3 methods of locating the function to be dispatched work and neither of them chains any exceptions when the function gets called

In order to get 100 % coverage of the `_dispatch` function, though, I would also need to test the negative case when no function can be located, however the exception raised in that case is a simple `Exception` which is very generic. What do you suggest to do about this? Should I ignore the fact that the exception is generic and still try to catch it and make sure that what I caught was an instance of `Exception` and not a subclass of it? Something else?

Thanks

Commit:
issue29615 (review): tests that ensure no exception chaining takes place when calling dispatched funcs

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,7 +343,53 @@ def run_server():
+        try:
+            dispatcher._dispatch(method='dispatched_func', params=exp_params)
+        except DispatchExc as e:
+            self.assertFalse(e.__cause__)
```
`assertIsNone()`

petr-motejlek replied:
Done

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,7 +343,53 @@ def run_server():
+        dispatcher = xmlrpc.server.SimpleXMLRPCDispatcher()
+        dispatcher.register_function(dispatched_func)
+        try:
+            dispatcher._dispatch(method='dispatched_func', params=exp_params)
```
I think it would be better to pass arguments as positional. It is shorter and more reliable. Parameter names is an implementation detail that we don't want to test here.

petr-motejlek replied:
Sure. I am sorry about this, but in my code I tend to use keywords all the time, it makes it so much easier to understand what the code does when you see the keywords :). Removed.

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,7 +343,53 @@ def run_server():
+        dispatcher.register_function(dispatched_func)
+        try:
+            dispatcher._dispatch(method='dispatched_func', params=exp_params)
+        except DispatchExc as e:
```
The test is passed if `DispatchExc` is not raised.

Use `assertRaises()` as a context manager.
```
with self.assertRaises(DispatchExc) as cm:
    dispatcher._dispatch('dispatched_func', exp_params)
self.assertIsNone(cm.exception.__cause__)
```

petr-motejlek replied:
OH, snap. I always forget that context variables are visible outside the `with` statement. Sure will do this, that's great.

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,6 +343,78 @@ def run_server():
 
+    def test_simple_xml_rpc_dispatcher_resolves_registered_func_and_does_not_chain_excs_when_calling_it(self):
+        exp_params = 1, 2, 3
+        dispatched = False
```
Checking that `dispatched` is set is redundant in these tests. It is obvious that it is set since the exception is raised after that. By removing `dispatched` you can save few lines of code.

petr-motejlek replied:
OK

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,6 +343,78 @@ def run_server():
+        exp_params = 1, 2, 3
+        dispatched = False
+
+        class DispatchExc(Exception):
```
`DispatchExc` is used in several tests. It is worth to define it at the module level.

petr-motejlek replied:
I hope it's ok to define it at a test case level for now. (`SimpleXMLRPCDispatcherTestCase.DispatcherExc` that is)

serhiy-storchaka replied:
You will need to specify long qualified name (`SimpleXMLRPCDispatcherTestCase.DispatcherExc`) or use outer `self` to access it.

petr-motejlek replied:
I know :)

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,6 +343,78 @@ def run_server():
+            """Raised inside the dispatched func"""
+
+        class DispatchedClass:
+            @staticmethod
```
Why static method? It works for instance methods too.

petr-motejlek replied:
No other reason in particular other than PyCharm suggesting to make it static as it did not depend on `self`. But I will get rid of the static declaration.

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,6 +343,78 @@ def run_server():
             self.assertEqual(p.method(), 5)
             self.assertEqual(p.method(), 5)
 
+    def test_simple_xml_rpc_dispatcher_resolves_registered_func_and_does_not_chain_excs_when_calling_it(self):
```
Oh, no, don't use such long test names! You can use docstrings for describing what tests do.

petr-motejlek replied:
OK. I will also create a separate test case SimpleXMLRPCDispatcherTestCase to hold these.

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,6 +343,78 @@ def run_server():
+        def dispatched_func(*params):
+            nonlocal dispatched
+            dispatched = True
+            self.assertEqual(params, exp_params)
```
If `assertEqual()` fails, does this failure correctly reported? I would pass params as arguments to `DispatchExc()` and check the `args` attribute of the catched exception.

petr-motejlek replied:
Sure, let me remove the delegation and assert in the test itself

serhiy-storchaka commented on Lib/xmlrpc/server.py:
```diff
@@ -392,31 +392,39 @@ def _dispatch(self, method, params):
+            return func(*params)
+
+        if self.instance is None:
             raise Exception('method "%s" is not supported' % method)
```
`raise Exception()` is repeated multiple times. I think it is enough to repeat it only two times. Once when `self.funcs[method]` is `None`, and other time at the end of this method, as in the previous version of your code.

petr-motejlek replied:
OK, right. That's true. Done.

serhiy-storchaka commented on Lib/xmlrpc/server.py:
```diff
@@ -392,31 +392,39 @@ def _dispatch(self, method, params):
+        if hasattr(self.instance, '_dispatch'):
+            # call the `_dispatch` method on the instance
+            return self.instance._dispatch(method, params)
+        else:
```
Since the other branch is ended by `return`, `else` is redundant.

petr-motejlek replied:
Fixed

serhiy-storchaka commented on Misc/NEWS:
```diff
@@ -10,6 +10,10 @@ What's New in Python 3.7.0 alpha 1?
 Core and Builtins
 -----------------
 
+- issue29615: xmlrpc.server.SimpleXMLRPCDispatcher.instance._dispatch function is no longer called in an except block
```
The line is too long. Add periods after ends of sentences. Use "bpo-29615" instead of "issue29615". This entry should be in the "Library" section rather of "Core and Builtins".

I think it would be better to reformulate the entry as "SimpleXMLRPCDispatcher no longer adds KeyError exception in the traceback of exception raised in the method of the registered instance." Or something like. Describe the problem that is solved, not the solution.

petr-motejlek replied:
Sure, reworded

serhiy-storchaka commented on Misc/NEWS:
```diff
@@ -10,6 +10,10 @@ What's New in Python 3.7.0 alpha 1?
 -----------------
 
+- issue29615: xmlrpc.server.SimpleXMLRPCDispatcher.instance._dispatch function is no longer called in an except block
+  Patch by Petr MOTEJLEK
```
Patch by Petr Motejlek.

petr-motejlek replied:
Sure

serhiy-storchaka commented on Misc/NEWS:
```diff
@@ -10,6 +10,10 @@ What's New in Python 3.7.0 alpha 1?
+- issue29615: xmlrpc.server.SimpleXMLRPCDispatcher.instance._dispatch function is no longer called in an except block
+  Patch by Petr MOTEJLEK
+
+- bpo-29607: Fix stack_effect computation for CALL_FUNCTION_EX. 
```
Unrelated change.

petr-motejlek replied:
Should be gone

serhiy-storchaka reviewed (changes requested):
Yes, I hope the next iteration will be the last or close to end. :-)

Commit:
issue29615 (review): created a separate SimpleXMLRPCDispatcherTestCase and renamed tests in it to have shorter names

Commit:
issue29615 (review): defines test case level DispatchExc and uses it, uses self.assertRaises to ensure it was raised

Commit:
issue29615 (review): proper assert test

Commit:
issue29615 (review): removed the assert delegation from DispatchExc and moved it higher

Commit:
issue29615 (review): removed keyword arguments when calling `_dispatch`

Commit:
issue29615 (review): no longer staticmethod

Commit:
issue29615 (review): removed unnecessary else block

Commit:
issue29615 (review): fewer `raise`s

Commit:
issue29615 (review): reformulated and moved to the appropriate place

Commit:
issue29615 (review): removed

petr-motejlek commented:
Hi @serhiy-storchaka , I think I addressed all your comments. Please check.

There is still the coverage issue, though. What should I do about checking the raised exception when no methods can be located, please?

> In order to get 100 % coverage of the _dispatch function, though, I would also need to test the negative case when no function can be located, however the exception raised in that case is a simple Exception which is very generic. What do you suggest to do about this? Should I ignore the fact that the exception is generic and still try to catch it and make sure that what I caught was an instance of Exception and not a subclass of it? Something else?

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,77 +343,71 @@ def run_server():
-            self.assertFalse(e.__cause__)
-            self.assertFalse(e.__context__)
-        self.assertTrue(dispatched)
+        with self.assertRaises(SimpleXMLRPCDispatcherTestCase.DispatchExc) as exc_ctx:
```
The line if too long. Why not define `DispatchExc` at module level?

petr-motejlek replied:
I used `self` for now. I don't want to define it at module level just yet. There is nothing else using it. That kind of refactoring, in my book, should happen when somebody introduces another class (test case) that would use it

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,77 +343,71 @@ def run_server():
+    class DispatchExc(Exception):
+        """Raised inside the dispatched functions when checking for chained exceptions"""
+
+        def __init__(self, *args, method=None, params=None, **kwargs):
```
Since positional arguments are saved as the `args` attribute, I think the `__init__` method is redundant. Just check that `exc_ctx.exception.args` is a tuple of passed arguments (`(exp_params,)` or `('dispatched_func', exp_params)`). The simpler code the easier it for reading.

petr-motejlek replied:
I really do prefer keyword arguments and keyword structure, but hey, I learned something new about `Exception.args` and these are just tests and the code that instantiates the exception and the code that processes it are very close, so it is ok

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,77 +343,71 @@ def run_server():
-            """Raised inside the dispatched func"""
+class SimpleXMLRPCDispatcherTestCase(unittest.TestCase):
+    class DispatchExc(Exception):
+        """Raised inside the dispatched functions when checking for chained exceptions"""
```
The line is too long. Don't exceed the PEP 8 limit.

petr-motejlek replied:
79 chars, ok

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,77 +343,71 @@ def run_server():
+            self.params = params
+
+    def test_call_registered_func(self):
+        """Attempts to resolve an explicitly registered function and call it, making sure any exception raised
```
The line is too long. The docstring is output in tests log, don't be too verbose.

petr-motejlek replied:
Sure, 79 chars

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -1379,7 +1373,7 @@ def test_xmlrpcserver_has_use_builtin_types_flag(self):
 @support.reap_threads
 def test_main():
-    support.run_unittest(XMLRPCTestCase, HelperTestCase, DateTimeTestCase,
+    support.run_unittest(XMLRPCTestCase, SimpleXMLRPCDispatcherTestCase, HelperTestCase, DateTimeTestCase,
```
The line is too long.

petr-motejlek replied:
Of course, 79 chars.

serhiy-storchaka commented on Lib/xmlrpc/server.py:
```diff
@@ -402,13 +402,11 @@ def _dispatch(self, method, params):
                 raise Exception('method "%s" is not supported' % method)
```
For consistency with following code I would revert the condition (this is just a style nit):
```
if func is not None:
    return ...
raise ...
```

petr-motejlek replied:
Seems reasonable enough, ok

serhiy-storchaka commented:
I think it is enough to catch the general `Exception` and check that error message contains the name of the dispatched method. The type of the exception can be made more specific in future, and error message can be changed, but I hope that it will always contain the method name.

Commit:
issue29615 (review): uses `Exception.args`; shorter lines

Commit:
issue29615 (review): consisten use if `if` condition

Commit:
issue29615 (review): uses assertTupleEqual to compare tuples; tests the behavior when no function can be located by the dispatcher

Commit:
issue29615 (review): tests what happens when explicitly registered function is None

Commit:
issue29615 (review): checks what happens when a registered instance's attribute does not exist

petr-motejlek commented:
Ok, let's see now. I believe all comments have been addressed and there should be 100 % coverage on the updated _dispatch function

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -1307,10 +1398,11 @@ def test_xmlrpcserver_has_use_builtin_types_flag(self):
-            BinaryTestCase, FaultTestCase, UseBuiltinTypesTestCase,
-            SimpleServerTestCase, SimpleServerEncodingTestCase,
-            KeepaliveServerTestCase1, KeepaliveServerTestCase2,
+    support.run_unittest(XMLRPCTestCase, SimpleXMLRPCDispatcherTestCase,
```
I think it would be better to add new test at the end of the list of tests or at separate line. This would make the diff smaller and easier fore review.

petr-motejlek replied:
Sure

serhiy-storchaka commented on Misc/NEWS:
```diff
@@ -677,6 +677,10 @@ Library
 - Issue #24142: Reading a corrupt config file left configparser in an
   invalid state.  Original patch by Florian Höch.
 
+- bpo-29615: SimpleXMLRPCDispatcher no longer chains KeyError (or any other
```
Entries usually are added at the start of corresponding section.

petr-motejlek replied:
Sure

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,6 +343,97 @@ def run_server():
+        dispatcher.register_function(dispatched_func)
+        with self.assertRaises(self.DispatchExc) as exc_ctx:
+            dispatcher._dispatch('dispatched_func', exp_params)
+        self.assertTupleEqual(exc_ctx.exception.args, (exp_params,))
```
Just `assertEqual()` can be used. It calls `assertTupleEqual()` for tuples.

petr-motejlek replied:
Darn. And here I thought you would like this :D

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -343,6 +343,97 @@ def run_server():
+        chained exceptions"""
+
+    def test_call_registered_func(self):
+        """Calls explicitly registered function
```
Comments and docstrings that consist of multiple sentences should have periods at the end of sentences.

If you want the first line of the docstring be without a period, turn the rest of the docstring into a comment.
```python
        """Calls explicitly registered function"""
        # Makes sure any exception raised inside the function has no other
        # exception chained to it
```

petr-motejlek replied:
I will fix this, this seems reasonable

serhiy-storchaka reviewed (approved):
In general LGTM. Nice fix, nice tests, nice Misc/NEWS entry. I left few style comments, you can address them if you with, but this is not required.

Commit:
issue29615 (review): typography

Commit:
issue29615 (review): uses `assertEqual` instead of `assertTupleEqual`

Commit:
issue29615 (review): shorter diff

Commit:
Merge remote-tracking branch 'MASTER/master' into issue29615

Commit:
issue29615 (review): moving the NEWS item to the proper place

petr-motejlek commented:
Ok, @serhiy-storchaka . I think once the sanity checks pass, we are golden.

Thank you very much for the review and patience :)

serhiy-storchaka commented on Lib/test/test_xmlrpc.py:
```diff
@@ -350,10 +350,9 @@ class DispatchExc(Exception):
-        exception chained to it"""
+        """Calls explicitly registered function"""
+        # Makes sure any exception raised inside the function has no other
+        # exception chained to it"""
```
`"""` is not needed.

petr-motejlek replied:
Hasty typo. Removed.

Commit:
issue29615 (review): typo

serhiy-storchaka reviewed (approved):
Thank you for your contribution and patience to my endless nitpicks. ;-)

serhiy-storchaka commented:
I think it is worth to backport this to 3.6 and 3.5. Could you please do this @petr-motejlek?

petr-motejlek commented:
Will be my pleasure, @serhiy-storchaka. I wanted to do that in the first place, but did not know the proper procedure. What's that? Do I create separate PRs to the respective version branches?

serhiy-storchaka commented:
Yes, just create separate PRs. Add corresponding prefix ("[3.6]") and label ("cherry-pick for 3.6").

petr-motejlek commented:
Created 3.5 (https://github.com/python/cpython/pull/479) and 3.6 (https://github.com/python/cpython/pull/478) backport PRs.
